File: 3743.maximize-cyclic-partition-score.py
#
# @lc app=leetcode id=3743 lang=python3
#
# [3743] Maximize Cyclic Partition Score
#

# @lc code=start
class Solution:
    def maximumScore(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [[-1] * (k + 1) for _ in range(n)]  # DP table for memoization
        
        def max_score(start: int, remaining_k: int) -> int:
            if remaining_k == 0:
                return 0
            if dp[start][remaining_k] != -1:
                return dp[start][remaining_k]
            
            current_max, current_min = float('-inf'), float('inf')
            result = 0
            
            # Consider all end points for this partition starting from `start`
            for length in range(1, n + 1):
                idx = (start + length - 1) % n  # Wrap-around using modulo operation
                current_max = max(current_max, nums[idx])
                current_min = min(current_min, nums[idx])
                current_range = current_max - current_min
                score_with_partition = current_range + max_score((start + length) % n, remaining_k - 1)
                result = max(result, score_with_partition)
                if length == n:
                    break  # Avoid wrapping around more than once
            dp[start][remaining_k] = result
            return result
        
        return max(max_score(i, k) for i in range(n))
# @lc code=end

File: 3743.maximize-cyclic-partition-score.go
#
# @lc app=leetcode id=3743 lang=golang
#
# [3743] Maximize Cyclic Partition Score
#

# @lc code=start
func maximumScore(nums []int, k int) int64 {
    n := len(nums)
    extended := append(nums, nums...)
    dp := make([][]int64, k+1)
    for i := range dp {
        dp[i] = make([]int64, n*2)
    }
    
    for i := 0; i < n*2; i++ {
        for parts := 1; parts <= k; parts++ {
            if parts == 1 {
                maxNum := int64(extended[i])
                minNum := int64(extended[i])
                for j := i; j >= 0 && j > i-n; j-- {
                    maxNum = max(maxNum, int64(extended[j]))
                    minNum = min(minNum, int64(extended[j]))
                    dp[parts][i] = max(dp[parts][i], maxNum-minNum)
                }
            } else {
                maxNum := int64(extended[i])
                minNum := int64(extended[i])
                for j := i - 1; j >= 0 && j > i-n; j-- {
                    maxNum = max(maxNum, int64(extended[j]))
                    minNum = min(minNum, int64(extended[j]))
                    dp[parts][i] = max(dp[parts][i], dp[parts-1][j]+maxNum-minNum)
                }
            }
        }
    }
    
    result := int64(0)
    for i := n-1; i < n*2-1; i++ {
        result = max(result, dp[k][i])
    }
    return result
}
the
func max(a, b int64) int64 { if a > b { return a } else { return b } }
the
func min(a, b int64) int64 { if a < b { return a } else { return b } }
the 
dp code=end